### ISSUE TITLE
tf.math.l2_normalize with default axis normalizes per-last-axis instead of over the whole tensor like eager TensorFlow

### ISSUE THREAD

**Issue by laolvfan** [google-ai-edge/LiteRT#9212]
## Environment

- LiteRT: `ai-edge-litert 2.1.6`
- TensorFlow converter: `tf-nightly 2.22.0.dev20260808` (also reproduces bit-identically on `tf-nightly 2.22.0.dev20260726` and on TensorFlow 2.20.0)
- OS: Linux x86_64, CPU
- Python: 3.11

## Minimal reproduction

```python
import tensorflow as tf
from ai_edge_litert.interpreter import Interpreter, OpResolverType


class Model(tf.keras.Model):
    def call(self, x):
        return tf.math.l2_normalize(x)


x = tf.constant([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]], dtype=tf.float32)
model = Model()
expected = model(x).numpy()
flatbuffer = tf.lite.TFLiteConverter.from_keras_model(model).convert()


def run(**kwargs):
    interpreter = Interpreter(model_content=flatbuffer, **kwargs)
    input_detail = interpreter.get_input_details()[0]
    interpreter.resize_tensor_input(input_detail["index"], list(x.shape))
    interpreter.allocate_tensors()
    input_detail = interpreter.get_input_details()[0]
    interpreter.set_tensor(input_detail["index"], x.numpy())
    interpreter.invoke()
    return interpreter.get_tensor(interpreter.get_output_details()[0]["index"])


print("eager:", expected)
print("LiteRT default:", run())
print(
    "LiteRT without default delegates:",
    run(experimental_op_resolver_type=OpResolverType.BUILTIN_WITHOUT_DEFAULT_DELEGATES),
)
```

## Actual result

```text
eager: [[0.10482848 0.20965695 0.31448543]
 [0.4193139  0.5241424  0.62897086]]
LiteRT default: [[0.26726124 0.5345225  0.8017837 ]
 [0.45584232 0.5698029  0.68376344]]
LiteRT without default delegates: [[0.26726124 0.5345225  0.8017837 ]
 [0.45584232 0.5698029  0.68376344]]
```

`tf.math.l2_normalize(x, axis=None)` in eager TensorFlow normalizes over the
entire tensor (the L2 norm of all 6 elements, `sqrt(91) ≈ 9.539`, is used as
the denominator for every element). The converted LiteRT model instead
normalizes each row independently using only that row's own L2 norm
(`sqrt(14)` for row 0, `sqrt(77)` for row 1) — this is the behavior of
`axis=-1`, not `axis=None`. Both the default XNNPACK-delegated interpreter and
the interpreter without default delegates produce the same wrong result, so
this looks like a converter/graph-construction issue rather than a
delegate-specific kernel bug.

## Expected result

The converted model should match eager TensorFlow's whole-tensor
normalization for the default `axis=None`.

## Additional context

This is the same root cause as tensorflow/tensorflow#118181 and its duplicate
#120292 (both used a `[3, 1]` input, where the wrong per-row normalization of
a single-element "row" happens to produce exactly `[1.0, 1.0, 1.0]`). Both
were closed on 2026-06-09/2026-06-17 with a note that active TFLite
development has moved to LiteRT and to refile there if the issue persists —
they were not closed because the bug was fixed. I confirmed with the
reproducer above that this is still present as of
`tf-nightly 2.22.0.dev20260808` (bit-identical to `2.22.0.dev20260726` from
two weeks earlier), and could not find an existing LiteRT-repo report, so I'm
refiling it here per that guidance.

**Comment by laolvfan** [comment]
I dug into the converter source and found what looks like the exact bug. The
fusion pattern in `tflite/converter/transforms/optimize_patterns.td`
(`L2NormalizePattern1`..`6`) that rewrites `x * rsqrt(sum(square(x), axis))`
(or the `Div`/`Sqrt` variant) into `TFL_L2NormalizationOp` is gated by a
`L2NormValidReduceIndex` constraint, implemented as `L2NormalizeReduceAxis` in
`tflite/converter/transforms/optimize_pass.cc`:

```cpp
bool L2NormalizeReduceAxis(Value sq_op, DenseElementsAttr axis) {
  if (axis.getNumElements() == 0) {
    return false;
  }
  if (mlir::cast<ShapedType>(sq_op.getType()).getRank() - 1 ==
          *axis.getValues<int>().begin() ||
      *axis.getValues<int>().begin() == -1) {
    return true;
  }
  if (mlir::cast<ShapedType>(sq_op.getType()).getRank() !=
      axis.getNumElements()) {
    return false;
  }
  auto shape = mlir::cast<ShapedType>(sq_op.getType());
  SmallVector<int, 4> elems{axis.getValues<int>().begin(),
                            axis.getValues<int>().end()};
  for (int i = 0; i < shape.getRank(); ++i) {
    if (i != elems[i]) return false;
  }
  return true;
}
```

The last branch (reached when `axis.getNumElements() == rank`) accepts any
axis list that enumerates every dimension **in order**, i.e.
`axis == [0, 1, ..., rank-1]`. But `TFL_L2NormalizationOp` itself only ever
reduces along the trailing dimension (see `tflite/kernels/l2norm.cc` /
`tflite/kernels/internal/reference/l2normalization.h`, which hardcode
`trailing_dim = input_shape.DimensionsCount() - 1` with no axis parameter at
all). For a rank-1 input, `axis == [0]` genuinely is "reduce the only/last
axis", so this branch is correct there. But for any input with rank > 1, an
axis list `[0, 1, ..., rank-1]` means "reduce over *all* dimensions to a
scalar" — not "reduce over the trailing dimension only" — and those are only
equivalent when every non-trailing dimension has size 1. In my reproducer
`x` has shape `[2, 3]` (rank 2), `tf.math.l2_normalize(x, axis=None)` lowers
`reduce_sum`'s axis to the explicit list `[0, 1]`, this function wrongly
returns `true`, and the pattern fuses into `L2_NORMALIZATION`, which then
normalizes each row independently instead of the whole tensor — exactly the
mismatch I reported.

This looks like the "elems[i] == i for all i" branch was meant for some other
legitimately-equivalent case (I'd guess a rank-1-after-broadcast situation, or
was just a bug from the start) but as written it's unconditionally wrong for
rank > 1. A fix would need to either drop that branch, or add a rank == 1
guard so it can't fire when reducing more than one dimension actually changes
the result.

**Comment by shivvaam0001** [comment]
Hi @laolvfan,

Thank you for the reaching out to us. We have investigated this issue and reproduced the exact mismatch you described.
```
Eager:
[[0.10482848 0.20965695 0.31448543]
 [0.4193139  0.5241424  0.62897086]]

LiteRT:
[[0.26726124 0.5345225  0.8017837 ]
 [0.45584232 0.5698029  0.68376344]]
```
If you have a local fix ready or would like to implement these changes, please feel free to raise a Pull Request. We would be very happy to help review and coordinate the merge with you.

Thanks!

**Comment by laolvfan** [comment]
Hi, @shivvaam0001, sorry for my delay response.

I have open a PR #9464 related to this issue, please reopen this issue and 
review the PR code, thanks!
